### PR TITLE
HTBHF-1612 Added email verification and acceptance tests.

### PR DESCRIPTION
**Pull request by dwybourn**

### CODE REVIEW

File: src/test/acceptance/features/email-address.feature [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  Background:
+    Given I have entered my details up to the email address page
+
+  Scenario: Enter in a valid email address
+    When I enter a valid email address
+    Then I am shown the check details page
+
+  Scenario: Do not enter in an email address
+    When I do not enter an email address
+    Then I am informed that the email address is in the wrong format
```
Comment (mikeSmith08) [inline src/test/acceptance/features/email-address.feature]: Shouldn't this be `Then I am informed that I need to enter an email address`, hence different to the wrong format message?

File: src/test/acceptance/features/email-address.feature [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+  Scenario Outline: Fill email address with invalid format
```
Comment (gethyn1) [inline src/test/acceptance/features/email-address.feature]: Can this match the other scenarios `Fill email address with invalid format` -> `Enter in an invalid email address`

also maybe remove `in` for the scenarios: `Enter in a valid email address` -> `Enter a valid email address`

File: src/test/acceptance/features/email-address.feature (hunk 1 above)
Comment (YetAnotherMatt) [reply src/test/acceptance/features/email-address.feature]: How about if this is renamed 'I am informed that I must enter an email address in the correct format'

File: src/web/routes/application/email-address/validate.test.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    'email@domain.com',
+    'firstname.lastname@domain.com',
+    'email@subdomain.domain.com',
+    'firstname+lastname@domain.com',
+    '1234567890@domain.com',
+    'email@domain-one.com',
+    '_______@domain.com',
+    'email@domain.name',
+    'email@domain.co.jp',
+    'firstname-lastname@domain.com']
```
Comment (YetAnotherMatt) [inline src/web/routes/application/email-address/validate.test.js]: Can we add some non-ascii characters, e.g. cyrillic?

File: src/test/common/steps/email-address-steps.js [hunk 4]
```diff
@@ -1,7 +1,40 @@
...
+    pages.emailAddress.getEmailAddressLinkErrorCss(),
+    'Enter an email address in the correct format, like name@example.com')
+})
+
+Then(/^I see the email address (.*) in the textbox$/, async function (emailAddress) {
+  const enteredEmailAddress = await pages.emailAddress.getEmailAddressValue()
+  expect(enteredEmailAddress).to.be.equal(emailAddress)
+})
+
+const setNoValidate = () => {
```
Comment (mikeSmith08) [inline src/test/common/steps/email-address-steps.js]: Can we have this at the top of the file please?

File: src/web/routes/application/email-address/validate.test.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+  emailAddresses.forEach(emailAddress => {
+    const matches = EMAIL_ADDRESS_REGEX.test(emailAddress)
+    t.deepEqual(matches, true)
```
Comment (mikeSmith08) [inline src/web/routes/application/email-address/validate.test.js]: Do we need `deepEqual` here?

File: src/web/routes/application/email-address/validate.test.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  ]
+
+  emailAddresses.forEach(emailAddress => {
+    console.log(emailAddress)
+    const matches = EMAIL_ADDRESS_REGEX.test(emailAddress)
+    t.equal(matches, false)
+  })
+
+  t.end()
+})
```
Comment (gethyn1) [inline src/web/routes/application/email-address/validate.test.js]: I think we are missing a safe regex test here

File: src/web/server/locales/en/validation.json [hunk 7]
```diff
@@ -16,5 +16,6 @@
   "expectedDeliveryDateTooFarInFutureTitle": "The date you have entered is more than 8 months in the future. You must be at least 10 weeks pregnant to apply for yourself.",
   "expectedDeliveryDateTooFarInFuture": "If you have children under the age of 4, answer ‘no’ to this question to continue with this application on their behalf.",
   "missingPhoneNumber": "Enter a UK mobile number",
-  "invalidPhoneNumber": "Enter in a valid UK mobile number"
+  "invalidPhoneNumber": "Enter in a valid UK mobile number",
+  "invalidEmail": "Enter an email address in the correct format, like name@example.com"
```
Comment (mikeSmith08) [inline src/web/server/locales/en/validation.json]: Shouldn't we have `missingEmail` too as per my comment in the ATs above. Same for Welsh validation.json too.

File: src/web/routes/application/email-address/validate.test.js (hunk 6 above)
Comment (dwybourn) [reply src/web/routes/application/email-address/validate.test.js]: It doesn't pass the test, as it's a simple check for groups. But it is widely used and not reported to be unsafe.

File: src/test/common/steps/email-address-steps.js (hunk 4 above)
Comment (dwybourn) [reply src/test/common/steps/email-address-steps.js]: All other step files has the steps followed by methods.

File: src/web/routes/application/email-address/validate.test.js (hunk 3 above)
Comment (dwybourn) [reply src/web/routes/application/email-address/validate.test.js]: Done.

File: src/web/routes/application/email-address/validate.test.js (hunk 5 above)
Comment (dwybourn) [reply src/web/routes/application/email-address/validate.test.js]: Nope, replaced with equal.

File: src/test/acceptance/features/email-address.feature (hunk 1 above)
Comment (dwybourn) [reply src/test/acceptance/features/email-address.feature]: Renamed. Although we may need a discussion between the lack of consistency. Phone number has separate errors for missing and invalid number, nino and email only have one message to cover missing and invalid.

File: src/test/acceptance/features/email-address.feature (hunk 2 above)
Comment (dwybourn) [reply src/test/acceptance/features/email-address.feature]: The Scenario outline isn't matched to steps, just the when, given and then statements. 
Where we use 'Enter' in a step will say 'Enter in', so stuck with that for consistency.

File: src/web/server/locales/en/validation.json (hunk 7 above)
Comment (dwybourn) [reply src/web/server/locales/en/validation.json]: See top comment, we may need a missingEmail message or suffice with just one, such as in enter nino. Discussion needed.

Comment (YetAnotherMatt) [reply src/web/server/locales/en/validation.json]: This should be driven by Sam Hardie

Comment (dwybourn) [reply src/web/server/locales/en/validation.json]: As discussed, sticking with a single invalidEmail message.

**Pull request merged**